src/policy/concrete.rs

```diff
@@ -58,12 +59,12 @@ pub enum Policy<Pk: MiniscriptKey> {
     /// A HASH160 whose preimage must be provided to satisfy the descriptor.
     Hash160(Pk::Hash160),
     /// A list of sub-policies, all of which must be satisfied.
-    And(Vec<Policy<Pk>>),
+    And(Vec<Arc<Policy<Pk>>>),
     /// A list of sub-policies, one of which must be satisfied, along with
     /// relative probabilities for each one.
-    Or(Vec<(usize, Policy<Pk>)>),
+    Or(Vec<(usize, Arc<Policy<Pk>>)>),
```

Comment:
/me  gently banging head on table crying, "please god no, don't let this be a vector when it is only ever exactly 2 policies"

@apoelstra or @sanket1729  are `Or` and `And` policies only ever comprised of exactly 2 sub-policies?

Reply:
I think we might have such a restriction right now (you'll find some `panic`s enforcing it) but we intend to lift this restriction eventually.

The main issue is the computational complexity of compiling n-ary `or`s.

Reply:
> but we intend to lift this restriction eventually.

By we do you mean miniscript in general or rust-miniscript? From the miniscript docs it looks like its limited to 2 also but I couldn't tell from the syntax if that was guaranteed. Is it ambiguous for the same reason?

Reply:
Miniscript in general.

Reply:
Ah ok, so leave these as vecs then. Perhaps I'll add some comments. I literally nearly fell apart yesterday, after 3 days of mental modelling of vectors and tree iteration to find this "2 limit". I'm better today :)

Reply:
Glad to hear it :). Sure, some comments would be appreciated.

```
// Currently these vectors are limited to two elements. Eventually we would like to extend these
// to be n-ary, but first we need to decide on a game plan for how to efficiently compile n-ary
// disjunctions
```

or something.

Reply:
There are no such length restrictions on the `Policy::Threshold` so when `k == 1` we essentially have unbounded `Or` and when `k == v.len()` we have unbounded `And` - should we be restricting the length of the vector in these two cases also?

Reply:
FTR I'm playing around with creating a `Thresh<T>` type to maintain the invariants `v.len() > 0`, k != 0`, `k <= v.len()`.